synapse/storage/data_stores/main/e2e_room_keys.py

```diff
@@ -78,29 +38,59 @@ def set_e2e_room_key(self, user_id, version, room_id, session_id, room_key):
             StoreError
         """
 
-        yield self._simple_upsert(
+        yield self._simple_update_one(
             table="e2e_room_keys",
             keyvalues={
                 "user_id": user_id,
                 "version": version,
                 "room_id": room_id,
                 "session_id": session_id,
             },
-            values={
+            updatevalues={
                 "first_message_index": room_key["first_message_index"],
                 "forwarded_count": room_key["forwarded_count"],
                 "is_verified": room_key["is_verified"],
                 "session_data": json.dumps(room_key["session_data"]),
             },
-            lock=False,
+            desc="update_e2e_room_key",
         )
-        log_kv(
-            {
-                "message": "Set room key",
-                "room_id": room_id,
-                "session_id": session_id,
-                "room_key": room_key,
-            }
+
+    @defer.inlineCallbacks
+    def add_e2e_room_keys(self, user_id, version, room_keys):
+        """Bulk add room keys to a given backup.
+
+        Args:
+            user_id(str): the user whose backup we're adding to
+            version(str): the version ID of the backup for the set of keys we're adding to
+            room_keys(iterable[(str, str, dict)]): the keys to add, in the form
+                (roomID, sessionID, keyData)
```

Comment:
I don't love the fact that we have a storage-layer method making undocumented assumptions about the shape of the key data dict, but :man_shrugging: